src/algorithms/boundary.rs

```diff
@@ -0,0 +1,515 @@
+    }
+
+    /// Returns the port count in the boundary, both inputs and outputs.
+    pub fn num_ports(&self) -> usize {
+        self.inputs.len() + self.outputs.len()
+    }
+
+    /// Returns the [`BoundaryPort`] corresponding to a port index.
+    pub fn port(&self, port: PortIndex, direction: Direction) -> BoundaryPort {
+        let ports = match direction {
+            Direction::Incoming => &self.inputs,
+            Direction::Outgoing => &self.outputs,
+        };
+        let index = ports
+            .iter()
+            .position(|&input| input == port)
+            .expect("port not found in inputs");
+        BoundaryPort { index, direction }
+    }
+
+    /// Returns an iterator over the input or output ports in the boundary.
+    pub fn ports(&self, dir: Direction) -> impl Iterator<Item = BoundaryPort> {
+        let num_ports = match dir {
+            Direction::Incoming => self.inputs.len(),
+            Direction::Outgoing => self.outputs.len(),
+        };
+        (0..num_ports).map(move |index| BoundaryPort {
+            index,
+            direction: dir,
+        })
+    }
+
+    /// Returns an iterator over the input ports in the boundary.
+    #[inline]
+    pub fn inputs(&self) -> impl Iterator<Item = BoundaryPort> {
+        self.ports(Direction::Incoming)
+    }
+
+    /// Returns an iterator over the output ports in the boundary.
+    #[inline]
+    pub fn outputs(&self) -> impl Iterator<Item = BoundaryPort> {
+        self.ports(Direction::Outgoing)
+    }
+
+    /// Returns the [`PortIndex`] corresponding to a [`BoundaryPort`] in this boundary.
+    pub fn port_index(&self, port: &BoundaryPort) -> PortIndex {
+        match port.direction {
+            Direction::Incoming => self.inputs[port.index],
+            Direction::Outgoing => self.outputs[port.index],
+        }
+    }
+
+    /// Returns `true` if the other boundary contains the same amount of input
+    /// and output ports as this boundary.
+    ///
+    /// When two boundaries are compatible, [`BoundaryPort`]s that are valid in
+    /// one boundary are also valid in the other boundary.
+    pub fn is_compatible(&self, other: &Boundary) -> bool {
+        self.inputs.len() == other.inputs.len() && self.outputs.len() == other.outputs.len()
+    }
+
+    /// Computes a partial order between the ports of a boundary.
+    ///
+    /// Returns a map indicating for each input port in the boundary, the set of
+    /// output ports that can be reached from it in the graph.
+    ///
+    /// The `graph` parameter must be a valid graph for this boundary. When
+    /// multiple nodes not reachable from the boundary are present, consider
+    /// using a [`crate::view::Subgraph`] to restrict the traversal.
+    ///
+    /// # Complexity
+    ///
+    /// The complexity of this operation is `O(e log(n) + k*n)`, where `e` is
+    /// the number of links in the `graph`, is the number of nodes, and `k` is
+    /// the number of ports in the boundary.
```

Comment:
I am confused here whether the argument `graph: &impl LinkView` needs to be a graph of the bounded subgraph exactly, or it can be a larger graph that included the subgraph of interest.

The fact that on line 176 you filter for nodes with 0 incoming neighbours makes me think that this will only work if the bounded graph is the entire graph -- but in that case, isn't the boundary already specified by the set of disconnected ports?

From the user's perspective, I would find passing a graph for which the region of interest is only a subgraph a simpler interface to use -- otherwise I'd expect that in most cases the user will have to construct a `Subgraph` struct before being able to call this function.

If the intention is to let user pass large graphs of which only a subgraph is of interest, then doing a toposort of the entire graph on line 180 seems wasteful.

Let me know if I misunderstood something!